sympy/core/basic.py

```diff
@@ -93,14 +93,142 @@ class Basic(with_metaclass(ManagedProperties)):
+            # can optimise to avoid double sympification by returning None
+            # from _eval_args where possible.
+            if newargs is not None:
+                if cls._basic_sympifyargs:
+                    newargs = tuple(map(S, newargs))
+                args = newargs
+
+            # Now evaluate to something else
+            if evaluate is None:
+                evaluate = cls._basic_autodoit
+            if evaluate:
+                newobj = cls._eval_doit(*args)
+                if newobj is not None:
+                    return newobj
+
+        # Create the object and store the args. We get here either because:
+        # 1. The class uses _basic_version=1
+        # 2. The instance was created with evaluate=False
+        # 3. The object was unable to evaluate (_eval_doit returned None).
+        #
+        # For classes having _basic_version=1 the lines below are the whole of
+        # the Basic.__new__ method.
+
         obj = object.__new__(cls)
         obj._assumptions = cls.default_assumptions
         obj._mhash = None  # will be set by __hash__ method.
 
         obj._args = args  # all items in args must be Basic objects
         return obj
 
+    @classmethod
+    def _eval_args(cls, *args):
+        '''Handle optional/keyword arguments and validation for Basic.
+
+        This method performs the map
+
+            (*args, **kwargs) -> args
+
+        Where *args and **kwargs are the literal arguments passed when
+        creating the instance and args is a valid args tuple for the class.
+        Basic.__new__ will remove any evaluate= keyword argument but otherwise
+        pass everything through. If evaluate=False then the instance is
+        created with this args tuple so this method should raise if that will
+        not create a coherent object. When evaluate=True the args tuple will
+        be passed to _eval_doit which is responsible for evaluation and might
+        return an object with different args or of a different type.
+
+        This class method should be overridden by subclasses of Basic that use
+        _basic_version=2 to implement any logic for optional and keyword
+        arguments. The overridden implementation should:
+
+        - Return None if the args tuple is fine as it is.
+        - Return a new args tuple if necessary (because of e.g. optional
+          arguments).
+        - Raise TypeError/ValueError if the resulting args tuple will not
+          create a valid object.
+
+        The Basic._eval_args method does not accept any keyword arguments so
+        subclasses need to override this if they want to support that. The
+        method should then raise if any unrecognised keyword arguments are
+        found. After calling _eval_args Basic.__new__ will discard the
+        **kwargs that were passed so they can only have any effect if it
+        happens here.
+
+        Both *args and the values of **kwargs will have already been sympified
```

Comment:
If one have this as a `classmethod` it is possible to bypass the sympification. Is this really something we would like to enable?

Also, isn't it better that the methods returns a sympified answer? Apart from returning single numerical constants, I cannot really see that there will be any issues.

Reply:
Thanks for looking at this. I'll add a comment updating the status of this PR. I would like it to be reviewed but I also don't propose for it to be merged in its current state.

There is a class attribute `_basic_sympifyargs` which can be used to prevent sympification. This allows classes to opt out. I think it's good for the default to be automatic sympification though. Maybe I should discuss that in this docstring...

The `_basic_sympifyargs` option is used here in `ProductSet` to allow generators to be passed in as the first argument (although here I am deprecating that).